sdk/eventhub/azure-eventhub/azure/eventhub/_transport/_uamqp_transport.py

```diff
@@ -0,0 +1,650 @@
+
+        :param error: The error received in the send attempt.
+        :type error: Exception
+        :rtype: ~uamqp.errors.ErrorAction
+        """
+        if error.condition == b"com.microsoft:server-busy":
+            return errors.ErrorAction(retry=True, backoff=4)
+        if error.condition == b"com.microsoft:timeout":
+            return errors.ErrorAction(retry=True, backoff=2)
+        if error.condition == b"com.microsoft:operation-cancelled":
+            return errors.ErrorAction(retry=True)
+        if error.condition == b"com.microsoft:container-close":
+            return errors.ErrorAction(retry=True, backoff=4)
+        if error.condition in NO_RETRY_ERRORS:
+            return errors.ErrorAction(retry=False)
+        return errors.ErrorAction(retry=True)
+
+
+    class UamqpTransport(AmqpTransport):    # pylint: disable=too-many-public-methods
+        """
+        Class which defines uamqp-based methods used by the producer and consumer.
+        """
+        # define constants
+        MAX_FRAME_SIZE_BYTES = constants.MAX_FRAME_SIZE_BYTES
+        MAX_MESSAGE_LENGTH_BYTES = constants.MAX_MESSAGE_LENGTH_BYTES
+        TIMEOUT_FACTOR = 1000
+        CONNECTION_CLOSING_STATES: Tuple = (  # pylint:disable=protected-access
+                c_uamqp.ConnectionState.CLOSE_RCVD,  # pylint:disable=c-extension-no-member
+                c_uamqp.ConnectionState.CLOSE_SENT,  # pylint:disable=c-extension-no-member
+                c_uamqp.ConnectionState.DISCARDING,  # pylint:disable=c-extension-no-member
+                c_uamqp.ConnectionState.END,  # pylint:disable=c-extension-no-member
+            )
+
+        # define symbols
+        PRODUCT_SYMBOL = types.AMQPSymbol("product")
+        VERSION_SYMBOL = types.AMQPSymbol("version")
+        FRAMEWORK_SYMBOL = types.AMQPSymbol("framework")
+        PLATFORM_SYMBOL = types.AMQPSymbol("platform")
+        USER_AGENT_SYMBOL = types.AMQPSymbol("user-agent")
+        PROP_PARTITION_KEY_AMQP_SYMBOL = types.AMQPSymbol(PROP_PARTITION_KEY)
+
+        @staticmethod
+        def build_message(**kwargs):
+            """
+            Creates a uamqp.Message with given arguments.
+            :rtype: uamqp.Message
+            """
+            return Message(**kwargs)
+
+        @staticmethod
+        def build_batch_message(**kwargs):
+            """
+            Creates a uamqp.BatchMessage with given arguments.
+            :rtype: uamqp.BatchMessage
+            """
+            return BatchMessage(**kwargs)
+
+        @staticmethod
+        def to_outgoing_amqp_message(annotated_message):
+            """
+            Converts an AmqpAnnotatedMessage into an Amqp Message.
+            :param AmqpAnnotatedMessage annotated_message: AmqpAnnotatedMessage to convert.
+            :rtype: uamqp.Message
+            """
+            message_header = None
+            if annotated_message.header and any(annotated_message.header.values()):
```

Comment:
@annatisch - previously, we had:
```
            if annotated_message.header:
```
and were adding a header, even if there were no actual values in it. Updated this and found that pyamqp encoded messages are bigger, as you mentioned.